monitor/puller.go

```diff
@@ -0,0 +1,272 @@
+	if request.TriggerTxHash != "" {
+		log.Infoln("The intendSettle event has been synced on sgn")
+		return
+	}
+
+	disputeTimeout, err := m.ethClient.Ledger.GetDisputeTimeout(&bind.CallOpts{}, mainchain.Bytes2Cid(request.ChannelId))
+	if err != nil {
+		log.Errorln("GetDisputeTimeout err:", err)
+		return
+	}
+
+	request.DisputeTimeout = disputeTimeout.Uint64()
+	request.TriggerTxHash = rawLog.TxHash.Hex()
+	request.TriggerTxBlkNum = rawLog.BlockNumber
+	requestData := m.operator.CliCtx.Codec.MustMarshalBinaryBare(request)
+	msg := sync.NewMsgSubmitChange(sync.TriggerGuard, requestData, m.operator.Key.GetAddress())
+	log.Infof("submit change tx: trigger guard request %s", request)
+	m.operator.AddTxMsg(msg)
+}
+
+func (m *Monitor) syncValidator(address mainchain.Addr) {
+	ci, err := m.ethClient.DPoS.GetCandidateInfo(&bind.CallOpts{}, address)
+	if err != nil {
+		log.Errorln("Failed to query candidate info:", err)
+		return
+	}
+
+	commission, err := common.NewCommission(m.ethClient, ci.CommissionRate)
+	if err != nil {
+		log.Errorln("Failed to create new commission:", err)
+		return
+	}
+
+	vt := staking.Validator{
+		Description: staking.Description{
+			Identity: address.Hex(),
+		},
+		Tokens:     sdk.NewIntFromBigInt(ci.StakingPool).QuoRaw(common.TokenDec),
+		Status:     mainchain.ParseStatus(ci),
+		Commission: commission,
+	}
+
+	candidate, err := validator.CLIQueryCandidate(m.operator.CliCtx, validator.RouterKey, address.Hex())
+	if err != nil {
+		log.Errorln("sidechain query candidate err:", err)
+		return
+	}
+	v, err := validator.CLIQueryValidator(
+		m.operator.CliCtx, staking.RouterKey, candidate.Operator.String())
+	if err == nil {
+		if vt.Status.Equal(v.Status) && vt.Tokens.Equal(v.Tokens) &&
+			vt.Commission.CommissionRates.Rate.Equal(v.Commission.CommissionRates.Rate) {
+			log.Infof("no need to sync updated validator %x", address)
+			return
+		}
+	}
+
+	if m.ethClient.Address == address {
+		pk, err := sdk.GetPubKeyFromBech32(sdk.Bech32PubKeyTypeConsPub, viper.GetString(common.FlagSgnPubKey))
+		if err != nil {
+			log.Errorln("GetConsPubKeyBech32 err:", err)
+			return
+		}
+
+		vt.ConsPubKey = pk
+	}
+
+	validatorData := m.operator.CliCtx.Codec.MustMarshalBinaryBare(vt)
+	msg := sync.NewMsgSubmitChange(sync.SyncValidator, validatorData, m.operator.Key.GetAddress())
+	log.Infof("submit change tx: sync validator %x", address)
+	m.operator.AddTxMsg(msg)
+}
+
+func (m *Monitor) setTransactors() {
+	transactors, err := transactor.ParseTransactorAddrs(viper.GetStringSlice(common.FlagSgnTransactors))
+	if err != nil {
+		log.Errorln("parse transactors err", err)
+		return
+	}
+	setTransactorsMsg := validator.NewMsgSetTransactors(
+		mainchain.Addr2Hex(m.ethClient.Address),
+		transactors,
+		m.operator.Key.GetAddress(),
+	)
+	log.Infoln("set transactors", transactors)
+	m.operator.AddTxMsg(setTransactorsMsg)
+}
+
+func (m *Monitor) handleDPoSDelegate(delegate *mainchain.DPoSDelegate) {
```

Comment:
move this function to where it is called?

Reply:
done